File: internal/wal/wal_test.go
package wal

import (
	"path/filepath"
	"testing"

	"bigdb/internal/record"
)

func TestWALAppendAndReplay(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "wal.log")

	w, err := Open(path)
	if err != nil {
		t.Fatalf("open wal: %v", err)
	}
	defer w.Close()

	inputs := []record.Record{
		record.NewPut([]byte("a"), []byte("1"), 1),
		record.NewPut([]byte("b"), []byte("2"), 2),
		record.NewDelete([]byte("a"), 3),
	}

	for _, rec := range inputs {
		if err := w.Append(rec); err != nil {
			t.Fatalf("append failed: %v", err)
		}
	}

	var got []record.Record
	if err := w.Replay(func(r record.Record) error {
		got = append(got, r)
		return nil
	}); err != nil {
		t.Fatalf("replay failed: %v", err)
	}

	if len(got) != len(inputs) {
		t.Fatalf("replay length mismatch: got %d want %d", len(got), len(inputs))
	}

	for i := range inputs {
		if string(got[i].Key) != string(inputs[i].Key) {
			t.Fatalf("key mismatch at %d: got %q want %q", i, got[i].Key, inputs[i].Key)
		}
		if string(got[i].Value) != string(inputs[i].Value) {
			t.Fatalf("value mismatch at %d: got %q want %q", i, got[i].Value, inputs[i].Value)
		}
		if got[i].Timestamp != inputs[i].Timestamp {
			t.Fatalf("timestamp mismatch at %d: got %d want %d", i, got[i].Timestamp, inputs[i].Timestamp)
		}
		if got[i].IsDeleted != inputs[i].IsDeleted {
			t.Fatalf("deleted mismatch at %d: got %v want %v", i, got[i].IsDeleted, inputs[i].IsDeleted)
		}
	}
}

File: internal/wal/wal_test.cpp
#include "wal.h"
#include <gtest/gtest.h>
#include <filesystem>
#include "../record/record.h"

using namespace bigdb::wal;
using namespace bigdb::record;
namespace fs = std::filesystem;

class WALTest : public ::testing::Test {
protected:
    void SetUp() override {
        temp_dir = testing::TempDir() + "/wal_test_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count());
        fs::create_directories(temp_dir);
    }
    void TearDown() override {
        fs::remove_all(temp_dir);
    }
    std::string temp_dir;
};

TEST_F(WALTest, AppendAndReplay) {
    std::string path = temp_dir + "/wal.log";
    auto w = WAL::Open(path);
    ASSERT_NE(w, nullptr);

    std::vector<Record> inputs = {
        Record::new_put({'a'}, {'1'}, 1),
        Record::new_put({'b'}, {'2'}, 2),
        Record::new_delete({'a'}, 3)
    };

    for (const auto& rec : inputs) {
        EXPECT_NO_THROW(w->Append(rec));
    }

    std::vector<Record> got;
    EXPECT_NO_THROW(w->Replay([&](const Record& r) {
        got.push_back(r);
    }));

    ASSERT_EQ(got.size(), inputs.size());

    for (size_t i = 0; i < inputs.size(); i++) {
        EXPECT_EQ(got[i].key, inputs[i].key);
        EXPECT_EQ(got[i].value, inputs[i].value);
        EXPECT_EQ(got[i].timestamp, inputs[i].timestamp);
        EXPECT_EQ(got[i].is_deleted, inputs[i].is_deleted);
    }
}
